### PR TITLE
Use parallel push/pull in MultiAppTransfer

### PR DESCRIPTION
**Pull request by fdkong**
Refs #16879

### CODE REVIEW

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 1]
```diff
@@ -22,6 +22,10 @@
 #include "libmesh/parallel_algebra.h"
 #include "libmesh/dof_object.h"
 
+// TIMPI includes
+#include "timpi/communicator.h"
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:26]: All MooseObjects are ParallelObjects - you shouldn't ever need to include the comm includes

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 2]
```diff
@@ -345,38 +344,38 @@ MultiAppNearestNodeTransfer::execute()
...
-      _cached_dof_ids.resize(n_processors());
-    }
+    // Quadrature points I will receive from remote processors
+    std::map<processor_id_type, std::vector<Point>> incoming_qps;
+    auto qps_action_functor = [&incoming_qps](processor_id_type pid,
+                                              const std::vector<Point> & qps) {
+      // Quadrature points from processor 'pid'
+      auto & incoming_qps_from_pid = incoming_qps[pid];
+      // Store data for late use
+      std::copy(qps.begin(), qps.end(), std::back_inserter(incoming_qps_from_pid));
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:354]:
```suggestion
      incoming_qps_from_pid.reserve(incoming_qps_from_pid.size() + qps.size());
      std::copy(qps.begin(), qps.end(), std::back_inserter(incoming_qps_from_pid));
```

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 3]
```diff
@@ -345,38 +344,38 @@ MultiAppNearestNodeTransfer::execute()
...
       std::vector<Real> & outgoing_evals = processor_outgoing_evals[i_proc];
       // Resize this vector to two times the size of the incoming_qps
       // vector because we are going to store both the value from the nearest
       // local node *and* the distance between the incoming_qp and that node
       // for later comparison purposes.
-      outgoing_evals.resize(2 * incoming_qps.size());
+      outgoing_evals.resize(2 * qps.second.size());
 
-      for (unsigned int qp = 0; qp < incoming_qps.size(); qp++)
+      for (unsigned int qp = 0; qp < qps.second.size(); qp++)
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:376]:
```suggestion
      for (std::size_t qp = 0; qp < qps.second.size(); qp++)
```

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 4]
```diff
@@ -345,38 +344,38 @@ MultiAppNearestNodeTransfer::execute()
...
+    for (auto & qps : incoming_qps)
     {
-      // We either use our own outgoing_qps or receive them from
-      // another processor.
-      std::vector<Point> incoming_qps;
-      if (i_proc == processor_id())
-        incoming_qps = outgoing_qps[i_proc];
-      else
-        _communicator.receive(i_proc, incoming_qps);
+      processor_id_type i_proc = qps.first;
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:361]: Now that you're moving a map, you should switch `i_proc -> pid` for everything in this loop. I think that you can

```suggestion
      const processor_id_type pid = qps.first;
```

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 5]
```diff
@@ -456,27 +451,24 @@ MultiAppNearestNodeTransfer::execute()
...
-        _communicator.send(i_proc, outgoing_evals, send_evals[i_proc]);
     }
   }
 
+  auto evals_action_functor = [&incoming_evals](processor_id_type pid,
+                                                const std::vector<Real> & evals) {
+    // evals for processor 'pid'
+    auto & incoming_evals_for_pid = incoming_evals[pid];
+    // Copy evals for late use
+    std::copy(evals.begin(), evals.end(), std::back_inserter(incoming_evals_for_pid));
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:462]:
```suggestion
    incoming_evals_for_pid.reserve(incoming_evals_for_pid.size() + evals.size());
    std::copy(evals.begin(), evals.end(), std::back_inserter(incoming_evals_for_pid));
```

File: framework/src/transfers/MultiAppNearestNodeTransfer.C [hunk 6]
```diff
@@ -546,19 +538,22 @@ MultiAppNearestNodeTransfer::execute()
           // point. If there are multiple values from other processors
           // which are equidistant, the first one we check will "win".
           Real min_dist = std::numeric_limits<Real>::max();
-          for (unsigned int i_from = 0; i_from < incoming_evals.size(); i_from++)
+          for (auto & evals : incoming_evals)
           {
+            // processor Id
+            auto i_from = evals.first;
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppNearestNodeTransfer.C:544]: Same here - you should switch this to being named `pid` or something of the sort.

File: framework/src/transfers/MultiAppProjectionTransfer.C [hunk 7]
```diff
@@ -27,6 +27,10 @@
 #include "libmesh/string_to_enum.h"
 #include "libmesh/default_coupling.h"
 
+// TIMPI includes
+#include "timpi/communicator.h"
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppProjectionTransfer.C:31]: Should be able to remove

File: framework/src/transfers/MultiAppProjectionTransfer.C [hunk 8]
```diff
@@ -349,76 +340,71 @@ MultiAppProjectionTransfer::execute()
...
-    else
-    {
-      incoming_qps = _cached_qps[i_proc];
-    }
+    auto qps_action_functor = [&incoming_qps](processor_id_type pid,
+                                              const std::vector<Point> & qps) {
+      // Quadrature points from processor 'pid'
+      auto & incoming_qps_from_pid = incoming_qps[pid];
+      // Store data for late use
+      std::copy(qps.begin(), qps.end(), std::back_inserter(incoming_qps_from_pid));
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppProjectionTransfer.C:355]:
```suggestion
      incoming_qps_from_pid.reserve(incoming_qps_from_pid.size() + qps.size());
      std::copy(qps.begin(), qps.end(), std::back_inserter(incoming_qps_from_pid));
```

File: framework/src/transfers/MultiAppProjectionTransfer.C [hunk 9]
```diff
@@ -349,76 +340,71 @@ MultiAppProjectionTransfer::execute()
...
+  if (!_qps_cached)
+    _cached_qps = incoming_qps;
 
-    outgoing_evals[i_proc].resize(incoming_qps.size(), OutOfMeshValue);
-    if (_current_direction == FROM_MULTIAPP)
-      outgoing_ids[i_proc].resize(incoming_qps.size(), libMesh::invalid_uint);
-    for (unsigned int qp = 0; qp < incoming_qps.size(); qp++)
+  for (auto & qps : _cached_qps)
+  {
+    processor_id_type i_proc = qps.first;
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppProjectionTransfer.C:367]: Same here - name `pid` or something of the sort

File: framework/src/transfers/MultiAppProjectionTransfer.C [hunk 10]
```diff
@@ -349,76 +340,71 @@ MultiAppProjectionTransfer::execute()
...
+  // Values back from remote processors for my local quadrature points
+  std::map<processor_id_type, std::vector<std::pair<Real, unsigned int>>> incoming_evals_ids;
+
+  auto evals_action_functor =
+      [&incoming_evals_ids](processor_id_type pid,
+                            const std::vector<std::pair<Real, unsigned int>> & evals) {
+        // evals for processor 'pid'
+        auto & incoming_evals_ids_for_pid = incoming_evals_ids[pid];
+        // Copy evals for late use
+        std::copy(evals.begin(), evals.end(), std::back_inserter(incoming_evals_ids_for_pid));
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppProjectionTransfer.C:404]:
```suggestion
        incoming_evals_ids_for_pid.reserve(incoming_evals_ids_for_pid.size() + evals.size());
        std::copy(evals.begin(), evals.end(), std::back_inserter(incoming_evals_ids_for_pid));
```

File: framework/src/transfers/MultiAppProjectionTransfer.C [hunk 11]
```diff
@@ -446,8 +432,11 @@ MultiAppProjectionTransfer::execute()
         Point qpt = xyz[qp];
 
         unsigned int lowest_app_rank = libMesh::invalid_uint;
-        for (unsigned int i_proc = 0; i_proc < n_processors(); i_proc++)
+        for (auto & values_ids : incoming_evals_ids)
         {
+          // Current processor id
+          processor_id_type i_proc = values_ids.first;
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppProjectionTransfer.C:438]: Switch to `pid`

**Review by loganharbour — CHANGES_REQUESTED** [review]
All looks great - very beautiful code.

Small changes. Mostly involves reserving in vectors and changing variable names like `i_proc` to something more useful now that we don't index contiguously but use a map.

**Review by loganharbour — APPROVED** [review]
BeautifuL!

**Comment by loganharbour** [comment]
I added the parallel sweep tests. Would like to see them pass before getting this in.

**Comment by fdkong** [comment]
Good idea

**Comment by loganharbour** [comment]
I say we hold for a little bit until we figure out what to do about the failures.

**Comment by fdkong** [comment]
@rui-hu, Do you have suggestions on SAM failures here?

**Pull request merged**